### PR TITLE
Update PostLedgerEntryOperation retry logic

### PR DESCRIPTION
**Pull request by PallabPaul**
# Contributing to the Azure SDK

Fixes a bug for PostLedgerEntryOperation in both sync and async clients when using waitUntil.Complete. Azure Confidential Ledger generally has client IP-based stickiness so that users observe session consistency. Occasionally this affinity is lost unexpectedly, which can lead to errors when waiting for a commit (write is made to Node A but the read is on Node B, which doesn't know about the write yet). This can be disruptive, so we introduce a small change to retry some `HttpStatusCode.NotFound` errors for all scenarios.

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

**Review by lynshi — APPROVED** [review]

File: sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/PostLedgerEntryOperation.cs [hunk 1]
```diff
@@ -55,12 +55,37 @@ protected PostLedgerEntryOperation()
...
+                {
+                    break;
+                }
+
+                // Add a 0.5 second delay between retries.
+                int delayTimeMs = 500;
+                if (async) {
+                    await Task.Delay(delayTimeMs).ConfigureAwait(false);
+                } else {
+                    Thread.Sleep(delayTimeMs);
```
Comment (tg-msft) [inline sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/PostLedgerEntryOperation.cs:86]: Thread.Sleep does not play well with all SynchronizationContexts.  Please don't merge this as-is.  Ideally we'd be able to make this work with a custom DelayStrategy instead.

Comment (PallabPaul) [reply sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/PostLedgerEntryOperation.cs]: Based off of our email thread with @JoshLove-msft I considered both the `DelayStrategy` approach during the `WaitUntil.Completed` as well as the custom ResponseClassifier. I agree with Josh and think the custom ResponseClassifier is the better approach to not influence the polling pattern. I have added the custom ResponseClassifier to take into account 404 errors. Based off of the [default parameters of the RetryOptions class](https://github.com/Azure/azure-sdk-for-net/blob/ea7dc0d229a404ba85285ec63fe533a94650f051/sdk/core/Azure.Core/src/RetryOptions.cs#L40), a response resulting in 404 should now be retired 3 times with a delay of 0.8 seconds between each retry.

File: sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/ConfidentialLedgerClient.cs [hunk 2]
```diff
@@ -78,11 +78,19 @@ internal ConfidentialLedgerClient(Uri ledgerEndpoint, TokenCredential credential
                     Array.Empty<HttpPipelinePolicy>() :
                     new HttpPipelinePolicy[] { new BearerTokenAuthenticationPolicy(_tokenCredential, AuthorizationScopes) },
                 transportOptions,
-                new ResponseClassifier());
+                new ConfidentialLedgerResponseClassifier());
```
Comment (JoshLove-msft) [inline sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/ConfidentialLedgerClient.cs:81]: Does this client have other uses besides being the underlying client for the PostLedgerEntryOperation? If so, do we want this new behavior for all scenarios, or does it need to be scoped to the specific scenario?

Comment (PallabPaul) [reply sdk/confidentialledger/Azure.Security.ConfidentialLedger/src/ConfidentialLedgerClient.cs:81]: Yes this client does have other underlying uses including Getting Entries, User Data and other information. We observed this bug in the `PostLedgerEntryOperationAsync` method but it may occur in various other methods where our client is reaching out to our service and reading from a different node that it has written to. I've updated the changelog to mention that we are bringing this change to all scenarios.

**Review by JoshLove-msft — APPROVED** [review]

File: sdk/confidentialledger/Azure.Security.ConfidentialLedger/CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,12 @@
 # Release History
 
-## 1.2.0-beta.1 (Unreleased)
+## 1.2.0 (2023-08-03)
+
+### Bugs Fixed
+
+- Create a custom ResponseClassifier to allow `HttpStatusCode.NotFound` to be retriable. These errors may occur in scenarios where there is an unexpected loss of session stickiness when the connected node changes and transactions have not been fully replicated.
```
Comment (christothes) [inline sdk/confidentialledger/Azure.Security.ConfidentialLedger/CHANGELOG.md:7]: I would reword this to a customer focused scenario. For example:

"Service calls that result in a `HttpStatusCode.NotFound` status will now be retried by default. This is to handle scenarios where ...."

Comment (PallabPaul) [reply sdk/confidentialledger/Azure.Security.ConfidentialLedger/CHANGELOG.md]: Updated the wording.

**Review by christothes — APPROVED** [review]
